File: js/__tests__/testCoreOps.js
var Interpreter = require('/Users/garidor1/Desktop/planout/js/es6/interpreter');
/*
function runConfig(config, init={}) {
	var interpreter = new Interpreter(config, 'test_salt', init);
	return interpreter.get_params();
}

function run_config_single(config) {
	var x_config = {'op': 'set', 'var': 'x', 'value': config};
	return runConfig(x_config)['x'];
}

describe ("Test core operators", function() {
	it('should set appropriately', function() {
		var c = {'op': 'set', 'value': 'x_val', 'var': 'x'}
		var d = runConfig(c);
		expect(d).toEqual({ 'x': 'x_val'});
	});

	it('should work with seq', function() {
		var config = {'op': 'seq', 'seq': [
            {'op': 'set', 'value': 'x_val', 'var': 'x'},
            {'op': 'set', 'value': 'y_val', 'var': 'y'}
        ]};
        var d = runConfig(config)
        expect(d).toEqual({'x': 'x_val', 'y': 'y_val'});
	});

	it('should work with arr', function() {
		var arr = [4, 5, 'a']
        var a = run_config_single({'op': 'array', 'values': arr})
        expect(arr).toEqual(a);
	});

	it('should work with get', function() {
		var d = runConfig({
            'op': 'seq',
            'seq': [
                {'op': 'set', 'var': 'x', 'value': 'x_val'},
                {'op': 'set', 'var': 'y', 'value': {'op': 'get', 'var': 'x'}}
            ]
        });
        expect({'x': 'x_val', 'y': 'x_val'}).toEqual(d);
	});

	it('should work with cond', function() {
		var getInput = function(i, r) {
			return {'op': 'equals', 'left': i, 'right': r};
		};
		var testIf = function(i) {
			return runConfig({
				'op': 'cond',
           		'cond': [
	                {'if': getInput(i, 0),
	                 'then': {'op': 'set', 'var': 'x', 'value': 'x_0'}},
	                {'if': getInput(i, 1),
	                 'then': {'op': 'set', 'var': 'x', 'value': 'x_1'}}
            	]
			});
		};
		expect(testIf(0)).toEqual({ 'x': 'x_0'});
		expect(testIf(1)).toEqual({ 'x': 'x_1'});
	});

	it('should work with index', function() {
		var array_literal = [10, 20, 30];
		var obj_literal = {'a': 42, 'b': 43};

        var x = run_config_single(
            {'op': 'index', 'index': 0, 'base': array_literal}
        )
        expect(x).toEqual(10);
        
        x = run_config_single(
            {'op': 'index', 'index': 2, 'base': array_literal}
        )
        expect(x).toEqual(30);

        x = run_config_single(
            {'op': 'index', 'index': 'a', 'base': obj_literal}
        )
        expect(x).toEqual(42);

        x = run_config_single(
            {'op': 'index', 'index': 6, 'base': array_literal}
        )
        expect(x).toBe(undefined);

        x = run_config_single(
            {'op': 'index', 'index': 'c', 'base': obj_literal}
        )
        expect(x).toBe(undefined);

        x = run_config_single({
            'op': 'index',
            'index': 2,
            'base': {'op': 'array', 'values': array_literal}
        });
        expect(x).toEqual(30);
    });

	it('should work with length', function() {
		var arr = [0, 1, 2, 3, 4, 5];
        var length_test = run_config_single({'op': 'length', 'value': arr});
        expect(length_test).toEqual(arr.length);
        length_test = run_config_single({'op': 'length', 'value': []});
        expect(length_test).toEqual(0);
        length_test = run_config_single({'op': 'length', 'value':
                                        	{'op': 'array', 'values': arr}
                                        });
       	expect(length_test).toEqual(arr.length);
	});

	it('should work with not', function() {
		var x = run_config_single({'op': 'not', 'value': 0})
        expect(x).toBe(true);

        x = run_config_single({'op': 'not', 'value': false});
        expect(x).toBe(true);

        x = run_config_single({'op': 'not', 'value': 1})
        expect(x).toBe(false);

        x = run_config_single({'op': 'not', 'value': true});
        expect(x).toBe(false);
	});

	it('should work with or', function() {
		var x = run_config_single({
            'op': 'or',
            'values': [0, 0, 0]})
        expect(x).toBe(false);

        x = run_config_single({
            'op': 'or',
            'values': [0, 0, 1]})
        expect(x).toBe(true);

        x = run_config_single({
            'op': 'or',
            'values': [false, true, false]})
        expect(x).toBe(true);
	});    

	it('should work with and', function() {
		var x = run_config_single({
            'op': 'and',
            'values': [1, 1, 0]})
        expect(x).toEqual(false);

        x = run_config_single({
            'op': 'and',
            'values': [0, 0, 1]})
       expect(x).toBe(false);

        x = run_config_single({
            'op': 'and',
            'values': [true, true, true]})
        expect(x).toBe(true);
	});

	it('should work with commutative operators', function() {
        var arr = [33, 7, 18, 21, -3];

        var min_test = run_config_single({'op': 'min', 'values': arr});
        expect(min_test).toEqual(-3);

        var max_test = run_config_single({'op': 'max', 'values': arr});
        expect(max_test).toEqual(33);

        var sum_test = run_config_single({'op': 'sum', 'values': arr});
        expect(sum_test).toEqual(76);

        var product_test = run_config_single({'op': 'product', 'values': arr});
        expect(product_test).toEqual(-261954);
	});

	it('should work with binary operators', function() {
        var eq = run_config_single({'op': 'equals', 'left': 1, 'right': 2});
        expect(eq).toEqual(1 == 2);

        eq = run_config_single({'op': 'equals', 'left': 2, 'right': 2});
        expect(eq).toEqual(2 == 2);
   
        var gt = run_config_single({'op': '>', 'left': 1, 'right': 2});
        expect(gt).toEqual(1 > 2);
        
        var lt = run_config_single({'op': '<', 'left': 1, 'right': 2});
        expect(lt).toEqual(1 < 2);
        
        var gte = run_config_single({'op': '>=', 'left': 2, 'right': 2});
        expect(gte).toEqual(2 >= 2);
        gte = run_config_single({'op': '>=', 'left': 1, 'right': 2});
        expect(gte).toEqual(1 >= 2);

        var lte = run_config_single({'op': '<=', 'left': 2, 'right': 2});
        expect(lte).toEqual(2 <= 2);

        var mod = run_config_single({'op': '%', 'left': 11, 'right': 3});
        expect(mod).toEqual(11 % 3);

        var div = run_config_single({'op': '/', 'left': 3, 'right': 4})
        expect(div).toEqual(0.75);
	});

	it('should work with return', function() {
		var return_runner = function(return_value) {
            var config = {
                "op": "seq",
                "seq": [
                    {
                      "op": "set",
                      "var": "x",
                      "value": 2
                    },
                    {
                        "op": "return",
                        "value": return_value
                    },
                    {
                        "op": "set",
                        "var": "y",
                        "value": 4
                    }
                ]
            };
	        var e = new Interpreter(config, 'test_salt');
            return e;
        };
        var i = return_runner(true);
        expect(i.get_params()).toEqual({'x': 2});
        expect(i.in_experiment()).toEqual(true);

        i = return_runner(42);
        expect(i.get_params()).toEqual({ 'x': 2});
        expect(i.in_experiment()).toEqual(true);

        i = return_runner(false);
        expect(i.get_params()).toEqual({ 'x': 2});
        expect(i.in_experiment()).toEqual(false);

        i = return_runner(0);
        expect(i.get_params()).toEqual({ 'x': 2});
        expect(i.in_experiment()).toEqual(false);
	});

});*/

File: python/planout/test/test_core_ops.py
# Copyright (c) 2014, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree. An additional grant
# of patent rights can be found in the PATENTS file in the same directory.

import unittest
import six

from planout.interpreter import (
    Interpreter,
)


class TestBasicOperators(unittest.TestCase):

    def runConfig(self, config, init={}):
        e = None
        e = Interpreter(config, 'test_salt', init)
        return e.get_params()

    def run_config_single(self, config):
        x_config = {'op': 'set', 'var': 'x', 'value': config}
        return self.runConfig(x_config)['x']

    def test_set(self):
        """Test setter"""
        # returns experiment object with probability p
        c = {'op': 'set', 'value': 'x_val', 'var': 'x'}
        d = self.runConfig(c)
        self.assertEquals({'x': 'x_val'}, d)

    def test_seq(self):
        """Test sequence"""
        config = {'op': 'seq', 'seq': [
            {'op': 'set', 'value': 'x_val', 'var': 'x'},
            {'op': 'set', 'value': 'y_val', 'var': 'y'}
        ]}
        d = self.runConfig(config)
        self.assertEquals({'x': 'x_val', 'y': 'y_val'}, d)

    def test_array(self):
        arr = [4, 5, 'a']
        a = self.run_config_single({'op': 'array', 'values': arr})
        self.assertEquals(arr, a)

    def test_map(self):
        my_map = {'a': 2, 'b': 'c', 'd': False}
        m = self.run_config_single({'op': 'map', 'a': 2, 'b': 'c', 'd': False})
        self.assertEquals(my_map, m)
        
        my_map = {}
        m = self.run_config_single({'op': 'map'})
        self.assertEquals(my_map, m)

    def test_cond(self):
        getInput = lambda i, r: {'op': 'equals', 'left': i, 'right': r}
        testIf = lambda i: self.runConfig({
            'op': 'cond',
            'cond': [
                {'if': getInput(i, 0),
                 'then': {'op': 'set', 'var': 'x', 'value': 'x_0'}},
                {'if': getInput(i, 1),
                 'then': {'op': 'set', 'var': 'x', 'value': 'x_1'}}
            ]
        })
        self.assertEquals({'x': 'x_0'}, testIf(0))
        self.assertEquals({'x': 'x_1'}, testIf(1))

    def test_get(self):
        d = self.runConfig({
            'op': 'seq',
            'seq': [
                {'op': 'set', 'var': 'x', 'value': 'x_val'},
                {'op': 'set', 'var': 'y', 'value': {'op': 'get', 'var': 'x'}}
            ]
        })
        self.assertEquals({'x': 'x_val', 'y': 'x_val'}, d)

    def test_index(self):
        array_literal = [10, 20, 30]
        dict_literal = {'a': 42, 'b': 43}

        # basic indexing works with array literals
        x = self.run_config_single(
            {'op': 'index', 'index': 0, 'base': array_literal}
        )
        self.assertEquals(x, 10)

        x = self.run_config_single(
            {'op': 'index', 'index': 2, 'base': array_literal}
        )
        self.assertEquals(x, 30)

        # basic indexing works with dictionary literals
        x = self.run_config_single(
            {'op': 'index', 'index': 'a', 'base': dict_literal}
        )
        self.assertEquals(x, 42)

        # invalid indexes are mapped to None
        x = self.run_config_single(
            {'op': 'index', 'index': 6, 'base': array_literal}
        )
        self.assertEquals(x, None)

        # invalid indexes are mapped to None
        x = self.run_config_single(
            {'op': 'index', 'index': 'c', 'base': dict_literal}
        )
        self.assertEquals(x, None)

        # non literals also work
        x = self.run_config_single({
            'op': 'index',
            'index': 2,
            'base': {'op': 'array', 'values': array_literal}
        })
        self.assertEquals(x, 30)

    def test_coalesce(self):
        x = self.run_config_single({'op': 'coalesce', 'values': [None]})
        self.assertEquals(x, None)

        x = self.run_config_single(
            {'op': 'coalesce', 'values': [None, 42, None]})
        self.assertEquals(x, 42)

        x = self.run_config_single(
            {'op': 'coalesce', 'values': [None, None, 43]})
        self.assertEquals(x, 43)

    def test_length(self):
        arr = list(range(5))
        length_test = self.run_config_single({'op': 'length', 'value': arr})
        self.assertEquals(len(arr), length_test)
        length_test = self.run_config_single({'op': 'length', 'value': []})
        self.assertEquals(0, length_test)
        length_test = self.run_config_single({'op': 'length', 'value':
                                              {'op': 'array', 'values': arr}
                                              })
        self.assertEquals(length_test, len(arr))

    def test_not(self):
        # test not
        x = self.run_config_single({'op': 'not', 'value': 0})
        self.assertEquals(True, x)
        x = self.run_config_single({'op': 'not', 'value': False})
        self.assertEquals(True, x)

        x = self.run_config_single({'op': 'not', 'value': 1})
        self.assertEquals(False, x)
        x = self.run_config_single({'op': 'not', 'value': True})
        self.assertEquals(False, x)

    def test_or(self):
        x = self.run_config_single({
            'op': 'or',
            'values': [0, 0, 0]})
        self.assertEquals(False, x)

        x = self.run_config_single({
            'op': 'or',
            'values': [0, 0, 1]})
        self.assertEquals(True, x)

        x = self.run_config_single({
            'op': 'or',
            'values': [False, True, False]})
        self.assertEquals(True, x)

    def test_and(self):
        x = self.run_config_single({
            'op': 'and',
            'values': [1, 1, 0]})
        self.assertEquals(False, x)

        x = self.run_config_single({
            'op': 'and',
            'values': [0, 0, 1]})
        self.assertEquals(False, x)

        x = self.run_config_single({
            'op': 'and',
            'values': [True, True, True]})
        self.assertEquals(True, x)

    def test_commutative(self):
        # test commutative arithmetic operators
        arr = [33, 7, 18, 21, -3]

        min_test = self.run_config_single({'op': 'min', 'values': arr})
        self.assertEquals(min(arr), min_test)

        max_test = self.run_config_single({'op': 'max', 'values': arr})
        self.assertEquals(max(arr), max_test)

        sum_test = self.run_config_single({'op': 'sum', 'values': arr})
        self.assertEquals(sum(arr), sum_test)

        product_test = self.run_config_single({'op': 'product', 'values': arr})
        self.assertEquals(six.moves.reduce(lambda x, y: x * y, arr), product_test)

    def test_binary_ops(self):
        eq = self.run_config_single({'op': 'equals', 'left': 1, 'right': 2})
        self.assertEquals(1 == 2, eq)
        eq = self.run_config_single({'op': 'equals', 'left': 2, 'right': 2})
        self.assertEquals(2 == 2, eq)
        gt = self.run_config_single({'op': '>', 'left': 1, 'right': 2})
        self.assertEquals(1 > 2, gt)
        lt = self.run_config_single({'op': '<', 'left': 1, 'right': 2})
        self.assertEquals(1 < 2, lt)
        gte = self.run_config_single({'op': '>=', 'left': 2, 'right': 2})
        self.assertEquals(2 >= 2, gte)
        gte = self.run_config_single({'op': '>=', 'left': 1, 'right': 2})
        self.assertEquals(1 >= 2, gte)
        lte = self.run_config_single({'op': '<=', 'left': 2, 'right': 2})
        self.assertEquals(2 <= 2, lte)
        mod = self.run_config_single({'op': '%', 'left': 11, 'right': 3})
        self.assertEquals(11 % 3, mod)
        div = self.run_config_single({'op': '/', 'left': 3, 'right': 4})
        self.assertEquals(0.75, div)

    def test_exp(self):
        # test exp operator
        x = self.run_config_single({'op': 'exp', 'value': 1})
        self.assertEqual(2.718281828459045, x)

        x = self.run_config_single({'op': 'exp', 'value': -0.5})
        self.assertEqual(0.6065306597126334, x)

        x = self.run_config_single({'op': 'exp', 'value': 0.123})
        self.assertEqual(1.1308844209474893, x)

        x = self.run_config_single({'op': 'exp', 'value': 1.88})
        self.assertEqual(6.553504862191148, x)

    def test_sqrt(self):
        # test sqrt operator
        x = self.run_config_single({'op': 'sqrt', 'value': 1})
        self.assertEqual(1, x)

        with self.assertRaises(ValueError):
            x = self.run_config_single({'op': 'sqrt', 'value': -0.5})

        x = self.run_config_single({'op': 'sqrt', 'value': 0.123})
        self.assertEqual(0.3507135583350036, x)

        x = self.run_config_single({'op': 'sqrt', 'value': 1.88})
        self.assertEqual(1.3711309200802089, x)

    def test_return(self):
        def return_runner(return_value):
            config = {
                "op": "seq",
                "seq": [
                    {
                      "op": "set",
                      "var": "x",
                      "value": 2
                    },
                    {
                        "op": "return",
                        "value": return_value
                    },
                    {
                        "op": "set",
                        "var": "y",
                        "value": 4
                    }
                ]
            }
            e = Interpreter(config, 'test_salt')
            return e

        i = return_runner(True)
        self.assertEquals({'x': 2}, i.get_params())
        self.assertEquals(True, i.in_experiment)
        i = return_runner(42)
        self.assertEquals({'x': 2}, i.get_params())
        self.assertEquals(True, i.in_experiment)
        i = return_runner(False)
        self.assertEquals({'x': 2}, i.get_params())
        self.assertEquals(False, i.in_experiment)
        i = return_runner(0)
        self.assertEquals({'x': 2}, i.get_params())
        self.assertEquals(False, i.in_experiment)

if __name__ == '__main__':
    unittest.main()
